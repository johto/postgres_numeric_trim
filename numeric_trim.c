#include "postgres.h"
#include "funcapi.h"
#include "fmgr.h"
#include "utils/builtins.h"
#include "utils/numeric.h"


PG_MODULE_MAGIC;

Datum numeric_trim(PG_FUNCTION_ARGS);
PG_FUNCTION_INFO_V1(numeric_trim);

Datum
numeric_trim(PG_FUNCTION_ARGS)
{
	Numeric num = PG_GETARG_NUMERIC(0);
	char *str;
	char *dot;
	char *last;
	Datum result;

	if (numeric_is_nan(num))
		PG_RETURN_NUMERIC(num);
	str = DatumGetCString(DirectFunctionCall1(numeric_out, PG_GETARG_DATUM(0)));
	dot = strchr(str, '.');
	if (dot == NULL)
	{
		pfree(str);
		PG_RETURN_NUMERIC(num);
	}
	else if (strchr(dot + 1, '.') != NULL)
	{
		/* shouldn't happen */
		pfree(str);
		PG_RETURN_NUMERIC(num);
	}
	last = strrchr(dot, '\0') - 1;
	while (*last == '0' && last > dot)
		last--;
	*(last + 1) = '\0';
	result = DirectFunctionCall3(numeric_in, CStringGetDatum(str), ObjectIdGetDatum(0), Int32GetDatum(0));
	pfree(str);
	PG_RETURN_DATUM(result);
}
