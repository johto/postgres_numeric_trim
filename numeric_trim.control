comment = 'numeric_trim'
default_version = '1.0'

relocatable = true
