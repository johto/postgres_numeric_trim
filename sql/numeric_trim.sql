CREATE EXTENSION numeric_trim;

SELECT numeric_trim(NULL);
SELECT numeric_trim('NaN');
SELECT numeric_trim('0');
SELECT numeric_trim('0.0');
SELECT numeric_trim('-0.0');
SELECT numeric_trim('1000.00');
SELECT numeric_trim('1000.0010');
SELECT numeric_trim('1.00000000000000000000000000000000000000000000000000000');
SELECT numeric_trim('1.000000000000000000000000000000000000000000000000000001');
SELECT numeric_trim('-1000.00');
SELECT numeric_trim('-1000.0010');
SELECT numeric_trim('-1.00000000000000000000000000000000000000000000000000000');
SELECT numeric_trim('-1.000000000000000000000000000000000000000000000000000001');
