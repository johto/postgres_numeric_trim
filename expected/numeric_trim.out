CREATE EXTENSION numeric_trim;
SELECT numeric_trim(NULL);
 numeric_trim 
--------------
             
(1 row)

SELECT numeric_trim('NaN');
 numeric_trim 
--------------
          NaN
(1 row)

SELECT numeric_trim('0');
 numeric_trim 
--------------
            0
(1 row)

SELECT numeric_trim('0.0');
 numeric_trim 
--------------
            0
(1 row)

SELECT numeric_trim('-0.0');
 numeric_trim 
--------------
            0
(1 row)

SELECT numeric_trim('1000.00');
 numeric_trim 
--------------
         1000
(1 row)

SELECT numeric_trim('1000.0010');
 numeric_trim 
--------------
     1000.001
(1 row)

SELECT numeric_trim('1.00000000000000000000000000000000000000000000000000000');
 numeric_trim 
--------------
            1
(1 row)

SELECT numeric_trim('1.000000000000000000000000000000000000000000000000000001');
                       numeric_trim                       
----------------------------------------------------------
 1.000000000000000000000000000000000000000000000000000001
(1 row)

SELECT numeric_trim('-1000.00');
 numeric_trim 
--------------
        -1000
(1 row)

SELECT numeric_trim('-1000.0010');
 numeric_trim 
--------------
    -1000.001
(1 row)

SELECT numeric_trim('-1.00000000000000000000000000000000000000000000000000000');
 numeric_trim 
--------------
           -1
(1 row)

SELECT numeric_trim('-1.000000000000000000000000000000000000000000000000000001');
                       numeric_trim                        
-----------------------------------------------------------
 -1.000000000000000000000000000000000000000000000000000001
(1 row)

